composer/algorithms/alibi/attention_surgery_functions/utils.py

```diff
@@ -0,0 +1,171 @@
+    """This decorator builds a registry that maps torch modules to their applicable SurgeryFunctionBuilder.
+
+    To accommodate the specifics of composer's model surgery, the register is used to create
+    a `Mapping[torch.nn.Module, SurgeryFunctionBuilder]`, where :func:`SurgeryFunctionBuilder` is any function that
+    takes a `max_sequence_length` argument and returns a :func:`composer.utils.model_surgery.ReplacementFunction`.
+
+    At runtime (see `../alibi.py`), ALiBi gets the model surgery policy mapping 
+    `Mapping[torch.nn.Module, ReplacementFunction]` by replacing each `SurgeryFunctionBuilder` with the 
+    `ReplacementFunction` it returns, then using the resulting `policies` mapping to perform
+    model surgery:
+
+    .. code-block:: python
+
+        from composer.algorithms.alibi.attention_surgery_functions import replacement_policy_mapping_builder
+        from composer.utils import model_surgery
+
+        policies = {
+            module: surgery_function_builder(max_sequence_length)
+            for module, surgery_function_builder in replacement_policy_mapping_builder.items()
+        }
+
+        module_surgery.replace_module_classes(model, policies=policies)
+
+
+    Implementation files (e.g., `../attention_surgery_functions/_gpt2.py`) populate the `replacement_policy_mapping_builder`
+    registry by defining instances of `SurgeryFunctionBuilder` functions and decorating them with
+    :func:`register_surgery_function_builder`. For example:
+
+    .. code-block:: python
+        from composer.algorithms.alibi.attention_surgery_functions.utils import register_surgery_function_builder
+        from transformers.models.gpt2.modeling_gpt2 import GPT2Attention
+
+        @register_surgery_function_builder(GPT2Attention)
+        def build_gpt2_attention_converter(max_sequence_length: int):
+            # Builds a function (`convert_attention`) that does model surgery any GPT2Attention modules in the model.
+            
+            def convert_attention(module: GPT2Attention, module_index: Optional[int] = None):
+                # Do surgery (change `module` or generate a new `module` instance to return)
+                # Note that this function can (and often should for ALiBi) depend on `max_sequence_length`
+                ...
+                return module
+            return convert_attention
+
+    In the above example, by decorating `build_gpt2_attention_converter` (which is an instances of a `SurgeryFunctionBuilder`
+    function) with `@register_surgery_function_builder(GPT2Attention)`, the ALiBi algorithm will now apply model surgery to any
+    instances of `GPT2Attention` within the model, and will apply surgery on those instances using the `convert_attention` function
+    returned by `build_gpt2_attention_converter`.
+
+    Note that `convert_attention` follows the specific signature of a `ReplacementFunction`, which has specific arguments. However,
```

Comment:
Is there model surgery documentation that can be linked to here?

Reply:
Added a reference to the `module_surgery.py` file